contracts/ButtonTokenWamplRouter.sol

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "./interfaces/IButtonWrapper.sol";
+import "./interfaces/IWAMPL.sol";
+import "./interfaces/IButtonToken.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "hardhat/console.sol";
+
+/**
+ * @dev Router to automatically wrap AMPL into WAMPL for ButtonToken actions
+ */
+contract ButtonTokenWamplRouter is ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    /// @notice ButtonToken underlying doesn't match wampl address
+    error InvalidButtonAsset();
+
+    /// @notice Zero amount provided
+    error ZeroAmount();
+
+    IWAMPL public immutable wampl;
+
+    constructor(address _wampl) {
+        wampl = IWAMPL(_wampl);
+    }
+
+    /**
+     * @notice Deposit the given amount of AMPL into the given WAMPL ButtonToken
+     *  Returns the output buttonTokens to the user.
+     * @dev Validates amount > 0
+     * @dev Validates buttonToken has WAMPL as underlying to prevent asset loss
+     * @dev buttonToken is kept as parameter to support multiple buttonToken implementations
+     *
+     * @param buttonToken the button token to deposit into
+     * @param amplAmount The amount of AMPL being deposited
+     * @return The amount of ButtonTokens created
+     */
+    function wamplWrapAndDeposit(IButtonToken buttonToken, uint256 amplAmount)
+        external
+        nonReentrant
+        returns (uint256)
+    {
+        if (amplAmount == 0) revert ZeroAmount();
+        if (buttonToken.underlying() != address(wampl)) revert InvalidButtonAsset();
+        // Transfer ampl to router
+        IERC20(wampl.underlying()).transferFrom(msg.sender, address(this), amplAmount);
+        // Approve ampl from router to wampl
+        IERC20(wampl.underlying()).safeApprove(address(wampl), amplAmount);
```

Comment:
Can we just set approval to infinite in a one off call in the constructor?
Would need to check if AMPL approval behaviour handled maxint specially where it doesn't decrement it as it gets used, else whilst very unlikely there's the possibility for router to brick if people somehow used up all maxint limit.